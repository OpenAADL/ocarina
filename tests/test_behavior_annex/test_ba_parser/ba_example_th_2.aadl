-- Execute conditions are used in the following example which merges sorted 
-- data received from two ports. Initially, arrival on either port can 
-- trigger a dispatch. Then arrival on one of the ports triggers the next 
-- dispatch. Each dispatch consumes one data, which is transferred from one 
-- of the two ports. Each transition to one of the two next states completes 
-- the execution of the thread, but its memory is supposed to be preserved so 
-- that it continues its execution at the next dispatch.

package BA_example2 public
with Base_Types;

thread merger
  features
    p1 : in event data port Base_Types::Integer;
    p2 : in event data port Base_Types::Integer;
    m : out event data port Base_Types::Integer;
end merger;

thread implementation merger.twopersistentstates 
subcomponents
  x1 : data Base_Types::Integer;
  x2 : data Base_Types::Integer;
annex behavior_specification {**
  states
    s0 : initial complete state;
    comp : state;
    next1, next2 : complete final state;
  transitions
    s0 -[ on dispatch p1 ]-> next2 { x1 := p1 };
    s0 -[ on dispatch p2 ]-> next1 { x2 := p2 };
    next1 -[ on dispatch p1 ]-> comp { x1 := p1 };  
    next2 -[ on dispatch p2 ]-> comp { x2 := p2 };
    comp -[ x1 < x2 ]-> next1 { m!(x1) };
    comp -[ x2 <= x1 ]-> next2 { m!(x2) };
**};
end merger.twopersistentstates;

-- Next version of the example utilizes the fact that an incoming port holds 
-- its previous value and only the port that triggers the dispatch has its 
-- input value frozen (updated).
thread implementation merger.portbased 
annex behavior_specification {**
  states
    s0 : initial complete state;
    comp : state;
    next1 : complete final state; 
    next2 : complete final state;
  transitions
    s0 -[ on dispatch p1 ]-> next2;
    s0 -[ on dispatch p2 ]-> next1;
    next1 -[ on dispatch p1 ]-> comp;  
    next2 -[ on dispatch p2 ]-> comp;
    comp -[ p1 < p2 ]-> next1 { m!(p1) };
    comp -[ p2 <= p1 ]-> next2 { m!(p2) };
**};
end merger.portbased;

end BA_example2;
