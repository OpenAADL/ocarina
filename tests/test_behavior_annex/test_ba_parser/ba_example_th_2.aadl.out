







property set Data_Model is
  Base_Type : list of classifier (
data)
    applies to (data);

  Code_Set : aadlinteger
    applies to (data);

  Data_Digits : aadlinteger
    applies to (data);

  Data_Scale : aadlinteger
    applies to (data);

  Data_Representation : enumeration (Array, Boolean, Character, Enum, Float, Fixed, Integer, String, Struct, Union)
    applies to (data);

  Dimension : list of aadlinteger
    applies to (data);

  Indefinite_Dimension : constant aadlinteger => -1;
  Infinite_Dimension : constant aadlinteger => -2;
  Element_Names : list of aadlstring
    applies to (data);

  Enumerators : list of aadlstring
    applies to (data);

  IEEE754_Precision : enumeration (Simple, Double)
    applies to (data);

  Initial_Value : list of aadlstring
    applies to (data, port, parameter);

  Integer_Range : range of aadlinteger
    applies to (data, port, parameter);

  Measurement_Unit : aadlstring
    applies to (data, port, parameter);

  Number_Representation : enumeration (Signed, Unsigned)
    applies to (data);

  Real_Range : range of aadlreal
    applies to (data, port, parameter);

  Representation : list of aadlstring
    applies to (data);

end Data_Model;

package Base_Types
public
  with data_model;

  data Boolean
  properties
    Data_Model::Data_Representation => Boolean;

  end Boolean;

  data Integer
  properties
    Data_Model::Data_Representation => Integer;

  end Integer;

  data Integer_8 extends Integer
  properties
    Data_Model::Number_Representation => Signed;
    Source_Data_Size => 1 Bytes;

  end Integer_8;

  data Integer_16 extends Integer
  properties
    Data_Model::Number_Representation => Signed;
    Source_Data_Size => 2 Bytes;

  end Integer_16;

  data Integer_32 extends Integer
  properties
    Data_Model::Number_Representation => Signed;
    Source_Data_Size => 4 Bytes;

  end Integer_32;

  data Integer_64 extends Integer
  properties
    Data_Model::Number_Representation => Signed;
    Source_Data_Size => 8 Bytes;

  end Integer_64;

  data Unsigned_8 extends Integer
  properties
    Data_Model::Number_Representation => Unsigned;
    Source_Data_Size => 1 Bytes;

  end Unsigned_8;

  data Unsigned_16 extends Integer
  properties
    Data_Model::Number_Representation => Unsigned;
    Source_Data_Size => 2 Bytes;

  end Unsigned_16;

  data Unsigned_32 extends Integer
  properties
    Data_Model::Number_Representation => Unsigned;
    Source_Data_Size => 4 Bytes;

  end Unsigned_32;

  data Unsigned_64 extends Integer
  properties
    Data_Model::Number_Representation => Unsigned;
    Source_Data_Size => 8 Bytes;

  end Unsigned_64;

  data Natural extends Integer
  properties
    Data_Model::Integer_Range => 0 .. Max_Target_Integer;

  end Natural;

  data Float
  properties
    Data_Model::Data_Representation => Float;

  end Float;

  data Float_32 extends Float
  properties
    Data_Model::IEEE754_Precision => Simple;
    Source_Data_Size => 4 Bytes;

  end Float_32;

  data Float_64 extends Float
  properties
    Data_Model::IEEE754_Precision => Double;
    Source_Data_Size => 8 Bytes;

  end Float_64;

  data Character
  properties
    Data_Model::Data_Representation => Character;

  end Character;

  data String
  properties
    Data_Model::Data_Representation => String;

  end String;

end Base_Types;

package BA_example2
public
  with Base_Types;

  thread merger
  features
    p1 : in event data port Base_Types::Integer;
    p2 : in event data port Base_Types::Integer;
    m : out event data port Base_Types::Integer;

  end merger;

  thread implementation merger.twopersistentstates
  subcomponents
    x1 : data Base_Types::Integer;
    x2 : data Base_Types::Integer;

    annex behavior_specification {**
    states
      s0 : initial complete state;
      comp : state;
      next1, next2 : complete final state;
    transitions      
      s0 -[on dispatch p1]-> next2 {
        x1  := p1
        };      
      s0 -[on dispatch p2]-> next1 {
        x2  := p2
        };      
      next1 -[on dispatch p1]-> comp {
        x1  := p1
        };      
      next2 -[on dispatch p2]-> comp {
        x2  := p2
        };      
      comp -[x1 < x2]-> next1 {
         m !(x1)
        };      
      comp -[x2 >= x1]-> next2 {
         m !(x2)
        };
    **};
  end merger.twopersistentstates;

  thread implementation merger.portbased
    annex behavior_specification {**
    states
      s0 : initial complete state;
      comp : state;
      next1 : complete final state;
      next2 : complete final state;
    transitions      
      s0 -[on dispatch p1]-> next2;      
      s0 -[on dispatch p2]-> next1;      
      next1 -[on dispatch p1]-> comp;      
      next2 -[on dispatch p2]-> comp;      
      comp -[p1 < p2]-> next1 {
         m !(p1)
        };      
      comp -[p2 >= p1]-> next2 {
         m !(p2)
        };
    **};
  end merger.portbased;

end BA_example2;

